lib/util/utils.ts

```diff
@@ -149,3 +156,41 @@ export function parseConnectionString<T>(connectionString: string, options?: Con
     };
   }, {} as any);
 }
+
+/**
+ * @ignore
+ * Describes the parameters to be provided to the function `emitEvent()`.
+ * @interface EmitParameters
+ */
+export interface EmitParameters {
+  rheaContext: RheaEventContext;
+  emitter: Link | Session | Connection;
+  eventName: string;
+  connectionId: string;
+  emitterType: "sender" | "receiver" | "session" | "connection";
+}
+
+/**
+ * @ignore
+ * Emits an event.
+ * @param params parameters needed to emit an event from one of the rhea-promise objects.
+ * @returns void
+ */
+export function emitEvent(params: EmitParameters): void {
+  const emit = () => {
+    log[params.emitterType]("[%s] %s got event: '%s'. Re-emitting the translated context.",
+      params.connectionId, params.emitterType, params.eventName);
+    params.emitter.emit(params.eventName,
+      EventContext.translate(params.rheaContext, params.emitter, params.eventName));
+  };
+  if (params.eventName.indexOf("error") !== -1) {
```

Comment:
@princjef - Currently we are emitting the error events with 1 tick delay. Should we do this only when an object (sender/receiver/session/connection) is being created? We can have a property called `isBeingCreated: boolean = false;` in a base class named `Entity`. All the classes would inherit from this base class. `isBeingCreated` would be set to true only while creating (i.e. establishing) a new object on the wire. It would be set to false, once it succeeded or failed. If the object is being created and we received an error event for that object, then 1 tick delay would be induced, else we would simply emit the event as it arrived. This would ensure we do not delay the notification unnecessarily.

Reply:
yeah you can add that if you want. specifically these are the cases:

 * creating but haven't resolved the promise yet - no delay needed
     * doesn't really matter because they can't attach a listener yet
 * creating, promise has been `resolve()`d but hasn't yet resolved to caller - delay needed to make sure they can register the listener
     * one could argue you want to wait 2 ticks for this: 1 tick for the promise to resolve and 1 more to make sure that you end up on a different tick and don't have to worry about ordering within a tick
 * created (promise fully resolved) - no delay needed
     * you would need to have a variable for whether you've resolved and then another one for whether the necessary tick has passed

the main argument against adding it is just that it's more complexity that you have to deal with and more tracking variables

Reply:
That is a good point. Will tweak this later if needed. Will keep it simple for now, as discussed.